src/System.Net.Http/src/System/Net/Http/Managed/ConnectHelper.cs

```diff
@@ -105,16 +105,31 @@ public BuilderAndCancellationTokenSocketAsyncEventArgs(CancellationToken cancell
 
         public static async ValueTask<SslStream> EstablishSslConnectionAsync(HttpConnectionSettings settings, string host, HttpRequestMessage request, Stream stream, CancellationToken cancellationToken)
         {
-            RemoteCertificateValidationCallback callback = null;
-            if (settings._serverCertificateCustomValidationCallback != null)
+            // Create the options bag to use.  As we mutate it, we don't just use the shared instance.
+            SslClientAuthenticationOptions sslOptions = settings._sslOptions != null ?
+                settings._sslOptions.ShallowClone() :
+                new SslClientAuthenticationOptions();
+
+            // Use the specified host, unless the options explicitly overrides it.
+            if (sslOptions.TargetHost == null)
             {
-                callback = (object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
+                sslOptions.TargetHost = host;
+            }
+
+            // If there's a cert validation callback, wrap the original delegate in order to:
+            // 1) Change the sender to be the request message (expected by HttpClientHandler's delegate), and
+            // 2) Wrap any exceptions in an HttpRequestException.
+            RemoteCertificateValidationCallback callback = sslOptions.RemoteCertificateValidationCallback;
+            if (callback != null)
+            {
+                // Replace the delegate, which will call to the old one.
+                sslOptions.RemoteCertificateValidationCallback = (object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
```

Comment:
Should we do this replacement magic only if the callback delegate is our special wrapper?  We're already testing for that in the settings handling.

Reply:
I wanted to keep the two layers entirely separate.  To do such detection, we'd need to move that type down into the managed handler, and then HttpClientHandler would need to know about its "internals".  It also seemed more consistent to make the sender always be the request message, regardless of whether accessed through HttpClientHandler or the managed handler.  I of course don't like that we're paying for it with additional allocations, though, so if you feel like it's worth the extra coupling, I can change it.

Reply:
Keeping them separate does seem like a good thing.  The additional allocation is probably not a huge deal.

I worry that we are committing ourselves to always having the HttpRequestMessage be the sender, even for non-wrapper cases; someone will notice this and then depend on it and we will never be able to get rid of this allocation, even when not using the wrapper model.  But I don't have a better solution here.